cobalt/browser/h5vcc_storage/h5vcc_storage_impl.cc

```diff
@@ -0,0 +1,196 @@
+#include "base/files/file_path.h"
+#include "base/files/file_util.h"
+#include "base/functional/bind.h"
+#include "base/functional/callback.h"
+#include "base/strings/stringprintf.h"
+#include "base/threading/thread_restrictions.h"
+#include "build/build_config.h"
+
+#if BUILDFLAG(IS_STARBOARD)
+#include "starboard/configuration_constants.h"
+#include "starboard/system.h"
+#endif
+
+#if BUILDFLAG(IS_ANDROID)
+#include "starboard/android/shared/starboard_bridge.h"
+
+using starboard::android::shared::StarboardBridge;
+#endif
+
+namespace h5vcc_storage {
+
+namespace {
+
+constexpr char kTestFileName[] = "cache_test_file.json";
+constexpr uint32_t kBufferSize = 16384;  // 16 KB
+
+base::FilePath GetCacheDirectory() {
+#if BUILDFLAG(IS_STARBOARD)
+  std::string path;
+  path.resize(kSbFileMaxPath);
+  if (!SbSystemGetPath(kSbSystemPathCacheDirectory, path.data(),
+                       kSbFileMaxPath)) {
+    NOTREACHED() << "Failed to get cache directory.";
+    return base::FilePath();
+  }
+  return base::FilePath(path.data());
+#elif BUILDFLAG(IS_ANDROID)
+  JNIEnv* env = base::android::AttachCurrentThread();
+  StarboardBridge* starboard_bridge = StarboardBridge::GetInstance();
+  return base::FilePath(starboard_bridge->GetCacheAbsolutePath(env));
+#else
+  NOTREACHED_NORETURN() << "Unsupported platform.";
+#endif
+}
+
+}  // namespace
+
+// TODO (b/395126160): refactor mojom implementation on Android
+H5vccStorageImpl::H5vccStorageImpl(
+    content::RenderFrameHost& render_frame_host,
+    mojo::PendingReceiver<mojom::H5vccStorage> receiver)
+    : content::DocumentService<mojom::H5vccStorage>(render_frame_host,
+                                                    std::move(receiver)) {}
+
+void H5vccStorageImpl::Create(
+    content::RenderFrameHost* render_frame_host,
+    mojo::PendingReceiver<mojom::H5vccStorage> receiver) {
+  new H5vccStorageImpl(*render_frame_host, std::move(receiver));
+}
+
+void H5vccStorageImpl::WriteTest(uint32_t test_size,
+                                 const std::string& test_string,
+                                 WriteTestCallback callback) {
+  base::ScopedAllowBlockingForTesting allow_blocking;
+  base::FilePath cache_dir = GetCacheDirectory();
+
+  LOG(WARNING) << "cache dir " << cache_dir;
+  base::DeletePathRecursively(cache_dir);
+  base::CreateDirectory(cache_dir);
+
+  auto response = mojom::WriteTestResponse::New();
+  response->error = "";
+  response->bytes_written = 0;
+
+  base::FilePath test_file_path = cache_dir.Append(kTestFileName);
+  base::File test_file(test_file_path,
+                       base::File::FLAG_OPEN_ALWAYS | base::File::FLAG_WRITE);
+
+  if (!test_file.IsValid()) {
+    response->error = base::StringPrintf("Error while opening ScopedFile: %s",
+                                         test_file_path.value().c_str());
+    std::move(callback).Run(std::move(response));
+    return;
+  }
+
+  // Repeatedly write test_string to test_size bytes of write_buffer.
```

Comment:
nit: Escape variable names in comments. It used to be `|test_string|` etc, now I think backticks are preferred.